Docs improvements: referencing

Bibo-Joshi opened the pull request:
* bump sphinx
* try to render base classes better
* use `sphinx-paramref` to make parameters referencable
* use nitpicky mode for sphinx
* add new CI task to build the docs

### Checklist for PRs

- [ ] Added `.. versionadded:: version`, `.. versionchanged:: version` or `.. deprecated:: version` to the docstrings for user facing changes (for methods/class descriptions, arguments and attributes)
- [ ] Created new or adapted existing unit tests
- [ ] Added myself alphabetically to `AUTHORS.rst` (optional)
- [ ] Added new classes & modules to the docs

harshil21 commented on telegram/_bot.py:
```diff
@@ -2819,7 +2826,7 @@ def set_webhook(
             url (:obj:`str`): HTTPS url to send updates to. Use an empty string to remove webhook
                 integration.
-            certificate (:obj:`filelike`): Upload your public key certificate so that the root
+            certificate (`filelike`): Upload your public key certificate so that the root
```
try 
```suggestion
            certificate (:term:`file-object`): Upload your public key certificate so that the root
```

it's definitely defined in python [glossary ](https://docs.python.org/3/glossary.html#term-file-object) or if `:term:` doesn't work, try `:any:` [sphinx says](https://www.sphinx-doc.org/en/master/usage/restructuredtext/roles.html#role-any) it should work...

harshil21 commented on telegram/_files/file.py:
```diff
@@ -102,37 +102,37 @@ def download(
 
-            * ``custom_path`` parameter now also accepts :obj:`pathlib.Path` as argument.
-            * Returns :obj:`pathlib.Path` object in cases where previously a :obj:`str` was
+            * ``custom_path`` parameter now also accepts :class:`pathlib.Path` as argument.
```
```suggestion
            * :paramref:`custom_path` parameter now also accepts :class:`pathlib.Path` as argument.
```

harshil21 commented on telegram/_files/sticker.py:
```diff
@@ -52,7 +52,7 @@ class Sticker(_BaseThumbedMedium):
         file_size (:obj:`int`, optional): File size in bytes.
         bot (:class:`telegram.Bot`, optional): The Bot to use for instance methods.
-        **kwargs (obj:`dict`): Arbitrary keyword arguments.
+        _kwargs (:obj:`dict`): Arbitrary keyword arguments.
```
ive seen this is many other places too, thought it's maybe ok to keep as is

Bibo-Joshi replied:
Yeah, I guess it doesn't matter too much since the parameter name is not actually used. This change was mainly due to me not realizing that the problem was the missing `:` in front of `obj` 😄

harshil21 commented on telegram/_passport/encryptedpassportelement.py:
```diff
@@ -40,7 +40,7 @@ class EncryptedPassportElement(TelegramObject):
 
     Objects of this class are comparable in terms of equality. Two objects of this class are
-    considered equal, if their :attr:`type`, :attr:`data`, :attr:`phone_number`, :attr:`email`,
+    considered equal, if their :attr:`type`, :attr:`data`, :paramref:`phone_number`, :attr:`email`,
```
um shouldn't the others also be paramrefs? Same in all files

Bibo-Joshi replied:
No, equality is determined by the attributes which (in general) may have little to do with the arguments. This change is probably due to me search-replacing a bit too thoroughly … Reverted

harshil21 commented on telegram/ext/_basepersistence.py:
```diff
@@ -195,7 +195,8 @@ def replace_bot(cls, obj: object) -> object:
         """
         Replaces all instances of :class:`telegram.Bot` that occur within the passed object with
         :attr:`REPLACED_BOT`. Currently, this handles objects of type ``list``, ``tuple``, ``set``,
-        ``frozenset``, ``dict``, ``defaultdict`` and objects that have a ``__dict__`` or
+        ``frozenset``, ``dict``, :class:`collections.defaultdict` and objects that have a
+        ``__dict__`` or
```
we can link the others too?

harshil21 commented on telegram/ext/_callbackqueryhandler.py:
```diff
@@ -44,11 +44,11 @@
+    Read the documentation of the :mod:`re` module for more information.
 
     Note:
         * If your bot allows arbitrary objects as ``callback_data``, it may happen that the
```
```suggestion
        * If your bot allows arbitrary objects as :paramref:`~telegram.InlineKeyboardButton.callback_data`, it may happen that the
```

harshil21 commented on telegram/ext/_choseninlineresulthandler.py:
```diff
@@ -47,7 +47,8 @@ class ChosenInlineResultHandler(Handler[Update, CCT]):
         run_async (:obj:`bool`): Determines whether the callback will run asynchronously.
             Defaults to :obj:`False`.
-        pattern (:obj:`str` | `Pattern`, optional): Regex pattern. If not :obj:`None`, ``re.match``
+        pattern (:obj:`str` | `Pattern`, optional): Regex pattern. If not :obj:`None`,
```
```suggestion
        pattern (:obj:`str` | :attr:`Pattern.pattern`, optional): Regex pattern. If not :obj:`None`,
```
do this? from https://docs.python.org/3/library/re.html#re.Pattern.pattern

Bibo-Joshi replied:
I used `:func:`re.Pattern <re.compile>`` in the other places - this prints `re.Pattern` but points to `re.compile`. this made sense for me b/c that's how you create pattern objects …. Added it here as well

harshil21 commented on telegram/ext/_contexttypes.py:
```diff
@@ -47,6 +47,16 @@ class ContextTypes(Generic[CCT, UD, CD, BD]):
             (error-)handler callbacks and job callbacks. Defaults to :obj:`dict`. Must support
             instantiating without arguments.
 
+    Attributes:
+        context (:obj:`type`): The type of the ``context`` argument of all
+            (error-)handler callbacks and job callbacks.
+        bot_data (:obj:`type`): The type of ``context.bot_data`` of all
+            (error-)handler callbacks and job callbacks.
+        chat_data (:obj:`type`): The type of ``context.chat_data`` of all
+            (error-)handler callbacks and job callbacks.
+        user_data (:obj:`type`): The type of ``context.user_data`` of all
+            (error-)handler callbacks and job callbacks.
+
```
we can doc this in the actual properties themselves?

harshil21 commented on telegram/ext/_dispatcher.py:
```diff
@@ -102,13 +102,15 @@ class Dispatcher(Generic[BT, CCT, UD, CD, BD, JQ, PT]):
             ``@run_async`` decorator and :meth:`run_async`.
-        user_data (:obj:`defaultdict`): A dictionary handlers can use to store data for the user.
-        chat_data (:obj:`defaultdict`): A dictionary handlers can use to store data for the chat.
+        user_data (:class:`collections.defaultdict`): A dictionary handlers can use to store dat
```
```suggestion
        user_data (:class:`collections.defaultdict`): A dictionary handlers can use to store data
```
at this point only a spell check CI will save us 😆

Bibo-Joshi replied:
the horror 😬

harshil21 commented on telegram/ext/_inlinequeryhandler.py:
```diff
@@ -60,9 +60,9 @@ class InlineQueryHandler(Handler[Update, CCT]):
-        pattern (:obj:`str` | :obj:`Pattern`, optional): Regex pattern. If not :obj:`None`,
-            ``re.match`` is used on :attr:`telegram.InlineQuery.query` to determine if an update
-            should be handled by this handler.
+        pattern (:obj:`str` | :func:`re.Pattern <re.compile>`, optional): Regex pattern.
```
oh maybe you can use this in my earlier comment above

harshil21 reviewed (changes requested):
Nice job! There's probably more places we can use `paramref` but for now this is a good start.

Can you build docs for this branch? I want to verify the 'hover' of paramrefs work fine on mobile

Bibo-Joshi commented on docs/source/_static/sphinx_paramlinks_override.css:
```diff
@@ -0,0 +1,20 @@
+}
+
+a.paramlink:after{
+    content: " \f0c1";
```
It inserts the link symbol https://fontawesome.com/v5.15/icons/link, the one that shows up when hovering. by default only the paragraph symbol ¶ is inserted, which doesn't match the link symbol used for the methods & attributes

Bibo-Joshi commented on docs/source/_static/sphinx_paramlinks_override.css:
```diff
@@ -0,0 +1,20 @@
+a.paramlink {
+    /*Uncomment this to make the link symbols show only on hover*/
+    /*But this introduces an ugly gap between parameter name & type*/
+	/*visibility: hidden;*/
+    color: inherit;
+    /*set font size to 0 to hide the "¶"*/
+    font-size: 0;
+}
+
+/*This is here for the hover part described above*/
+p:hover > a.headerlink, p:hover > a.paramlink, li:hover > a.paramlink, li:hover > a.headerlink {
+   visibility: visible;
```
will check it out 👍 
making the whole name the link sounds like a nice idea - will try it out. would you drop the link symbol, then?

harshil21 replied:
yep. It'll look just like regular text, but on hover you'll be able to copy it (click to copy/click to go to address). The site you linked above is an example. When hovering over the unicode text, it shows 'copy unicode'. Clicking it copies it. We could do something like that 
![image](https://user-images.githubusercontent.com/37377066/144130211-2c9f34c2-a48f-4c70-ad4c-55cce3ac3da1.png)

Bibo-Joshi commented on telegram/_bot.py:
```diff
@@ -486,10 +487,10 @@ def delete_message(
             - Bots can delete outgoing messages in private chats, groups, and supergroups.
             - Bots can delete incoming messages in private chats.
-            - Bots granted :attr:`telegram.ChatMember.can_post_messages` permissions can delete
+            - Bots granted ``can_post_messages`` permissions can delete
               outgoing messages in channels.
             - If the bot is an administrator of a group, it can delete any message there.
-            - If the bot has :attr:`telegram.ChatMember.can_delete_messages` permission in a
+            - If the bot has ``can_delete_messages`` permission in a
```
oh, I overlooked that that's the only one with that attribute. will do 👌

Bibo-Joshi commented on telegram/_bot.py:
```diff
@@ -1713,10 +1716,11 @@ def send_venue(
+            * You can either supply :paramref:`venue`, or :paramref:`latitude`,
+              :paramref:`longitude`,
+              :paramref:title` and :paramref:address` and optionally :paramref:`foursquare_id` and
+              :paramref:`foursquare_typ` or optionally :paramref:`google_place_id` and
+              :paramref:`google_place_typ`.
```
no … maybe that doesn't work for the param refs because for those the links arent inserted by sphinx itself but by the extension 🤔

harshil21 replied:
hmm, maybe we can raise an issue/pr upstream then, hopefully its possible to warn.

Bibo-Joshi commented on docs/source/conf.py:
```diff
@@ -386,7 +391,58 @@ def autodoc_skip_member(app, what, name, obj, skip, options):
+        # Drop generics (at least for now)
+        if base.endswith("]"):
+            base = base.split("[", maxsplit=1)[0]
+            bases[idx] = f':class:`{base}`'
+
```
No, e.g. `telegram._handler.Handler[…]` should be replaced by `telegram.Handler` rather than just `telegram._handler.Handler` :)

Bibo-Joshi commented:
Thanks for the review! Thorough as usual :) Didn't look at the CSS stuff yet though

Built docs are at https://python-telegram-bot.readthedocs.io/en/docs-referencing/

Bibo-Joshi commented:
Openend a PR at sqlalchemyorg/sphinx-paramlinks#13 for better nitpicky support and making parameter names the links

harshil21 commented:
While at it, we can also use nitpicky mode by default in the Makefile

Bibo-Joshi commented:
I added `nitpicky = True` to `docs/source/conf.py`. that should suffice, I guess 🤔

harshil21 commented on .github/workflows/docs.yml:
```diff
@@ -0,0 +1,39 @@
+          python -W ignore -m pip install -r requirements-dev.txt
+          python -W ignore -m pip install -r docs/requirements-docs.txt
+      - name: Build docs
+        run: sphinx-build docs/source docs/build/html -W --keep-going
```
We can add the argument `j auto` here. https://www.sphinx-doc.org/en/master/man/sphinx-build.html#cmdoption-sphinx-build-j.

I'm adding it to the makefile in #2841

Bibo-Joshi replied:
"this option should be considered _experimental_" ... Does it speed up much for you?

harshil21 replied:
by a factor of 10x or more

Bibo-Joshi replied:
o.0 for me, it didn't speed up at all 😄 but if it helps for you, let's add it - we can revert if there are problems and the feature has been around for 3 major versions now so probably not _that_ experimental anymore 🤔

harshil21 replied:
it really speeds it up when you run `make -C docs clean` and then `make -C docs html`

Bibo-Joshi replied:
jup, tried that. but it's probably also dependent on how many other processes are running on the machine etc …

harshil21 commented on telegram/_bot.py:
```diff
@@ -3964,7 +3973,7 @@ def set_chat_permissions(
         Use this method to set default chat permissions for all members. The bot must be an
         administrator in the group or a supergroup for this to work and must have the
-        :attr:`telegram.ChatMember.can_restrict_members` admin rights.
+        ``can_restrict_members`` admin rights.
```
this was missed

harshil21 commented on telegram/_bot.py:
```diff
@@ -4505,8 +4514,8 @@ def pin_chat_message(
-        and must have the :attr:`telegram.ChatMember.can_pin_messages` admin right in a supergroup
-        or :attr:`telegram.ChatMember.can_edit_messages` admin right in a channel.
+        and must have the ``can_pin_messages`` admin right in a supergroup
+        or ``can_edit_messages`` admin right in a channel.
```
and here

harshil21 commented on telegram/_bot.py:
```diff
@@ -4549,8 +4558,8 @@ def unpin_chat_message(
-        and must have the :attr:`telegram.ChatMember.can_pin_messages` admin right in a
-        supergroup or :attr:`telegram.ChatMember.can_edit_messages` admin right in a channel.
+        and must have the ``can_pin_messages`` admin right in a
+        supergroup or ``can_edit_messages`` admin right in a channel.
```
here

harshil21 commented on telegram/_bot.py:
```diff
@@ -4589,8 +4598,8 @@ def unpin_all_chat_messages(
-        to work and must have the :attr:`telegram.ChatMember.can_pin_messages` admin right in a
-        supergroup or :attr:`telegram.ChatMember.can_edit_messages` admin right in a channel.
+        to work and must have the ``can_pin_messages`` admin right in a
+        supergroup or ``can_edit_messages`` admin right in a channel.
```
here

harshil21 reviewed (commented):
some missed ones from the earlier review:

harshil21 commented:
might be worth to check how this looks in the new furo theme. Also I'm okay with merging this for now, hard to keep track of what is fixed and what isn't

Bibo-Joshi commented:
From this thread, two PRs were made at sphinx-paramlinks:

1. allow to customize how the link for the parameters is displayed
2. make sphinx-paramlinks support the nitpicky mode

The first is already merged, but there's no new release yet. The second is not yet merged.
I've updated the dependency to install from the current commit on sphinx-pl/master, which inludes 1 and changed the style so that the parameter names are clickable as discussed above.
IMO the nitpicky support is not very urgent, which is why I didn't put my fork as dependency.
